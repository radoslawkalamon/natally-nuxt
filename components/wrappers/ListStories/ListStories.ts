import Vue from 'vue'
import { DTOMetaPostStoryConstructor, DTOMetaPostStory } from '@/utils/dto.meta.post.story'

export default Vue.extend({
  name: 'WrappersListStories',
  props: {
    limit: {
      type: Number,
      default: 0
    },
    randomize: {
      type: Boolean,
      default: false
    },
    without: {
      type: Array as Vue.PropType<string[]>,
      default: () => []
    }
  },
  data (): Record<string, DTOMetaPostStoryConstructor[]> {
    return {
      stories: []
    }
  },
  async fetch (): Promise<void> {
    this.stories = this.randomize
      ? await this.fetchMetaPostStoryRandom()
      : await this.fetchMetaPostStory()
  },
  computed: {
    fetchContentDirectory (): string {
      return '/opowiadania'
    },
    metaPostStoryConstructorKeys (): string[] {
      return ['audiobookId', 'createdAt', 'description', 'imageCover', 'imageCover2x', 'imageOpenGraph', 'path', 'timeReading', 'title', 'updatedAt']
    },
    metaPostStories (): DTOMetaPostStory[] {
      return this.stories.map((i: DTOMetaPostStoryConstructor) => new DTOMetaPostStory(i))
    }
  },
  methods: {
    async fetchMetaPostStory (): Promise<DTOMetaPostStoryConstructor[]> {
      const stories = await this.$content(this.fetchContentDirectory)
        .where({ path: { $nin: this.without } })
        .only(this.metaPostStoryConstructorKeys)
        .sortBy('createdAt', 'desc')
        .limit(this.limit)
        .fetch<DTOMetaPostStoryConstructor>()

      return !Array.isArray(stories)
        ? [stories]
        : stories
    },
    async fetchMetaPostStoryRandom (): Promise<DTOMetaPostStoryConstructor[]> {
      const stories = await this.$content(this.fetchContentDirectory)
        .where({ path: { $nin: this.without } })
        .only(this.metaPostStoryConstructorKeys)
        .fetch<DTOMetaPostStoryConstructor>()

      return !Array.isArray(stories)
        ? [stories]
        : stories
          .sort(() => 0.5 - Math.random())
          .slice(0, this.limit || undefined)
    }
  }
})
