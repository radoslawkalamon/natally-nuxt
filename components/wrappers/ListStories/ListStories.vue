<template>
  <div
    class="list-stories"
    data-test="wrappers-list-stories"
  >
    <slot
      v-if="$fetchState.pending || $fetchState.error"
      name="loader"
    />
    <slot
      v-else
      v-bind="{ metaPostStories }"
    />
  </div>
</template>

<script src="./ListStories.ts" lang="ts"></script>
