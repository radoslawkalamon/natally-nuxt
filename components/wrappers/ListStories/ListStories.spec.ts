import { merge } from 'lodash'
import { mount } from '@vue/test-utils'
import ListStories from './ListStories.vue'
import { MockNuxtContent } from '@/devtools/jest.mock.nuxt.content'

jest.mock('@/utils/dto.meta.post.story', () => ({
  DTOMetaPostStory: class MockEmptyClass {}
}))

const defaultOptionsFactory = (options?: object) => merge({
  mocks: {
    $content: MockNuxtContent(Array(10).fill({}))
  }
}, options)

describe('Wrappers / List Stories', (): void => {
  test('shall render', async (): Promise<void> => {
    const wrapper = mount(ListStories, defaultOptionsFactory())
    await (ListStories as any).options.fetch.call(wrapper.vm)
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('shall have slot default', async (): Promise<void> => {
    const wrapper = mount(ListStories, defaultOptionsFactory({
      slots: {
        default: 'This is slot "default" test.'
      }
    }))
    await (ListStories as any).options.fetch.call(wrapper.vm)
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('shall { metaPostStories } be Array<DTOMetaPostStory>', async (): Promise<void> => {
    const wrapper = mount(ListStories, defaultOptionsFactory())
    await (ListStories as any).options.fetch.call(wrapper.vm)
    const metaPostStoriesConstructorNames = (wrapper.vm as any).metaPostStories.map((e: any) => e.constructor.name)
    expect(metaPostStoriesConstructorNames.every((e: string) => e === 'MockEmptyClass')).toBeTruthy()
  })

  describe('Randomize: false', (): void => {
    test('shall { metaPostStories } be Array fetch output Object', async (): Promise<void> => {
      const wrapper = mount(ListStories, defaultOptionsFactory({
        mocks: {
          $content: MockNuxtContent({})
        }
      }))
      await (ListStories as any).options.fetch.call(wrapper.vm)
      expect(Array.isArray((wrapper.vm as any).metaPostStories)).toBeTruthy()
    })

    test('shall $content.limit() be called with 3 when limit = 3', async (): Promise<void> => {
      const mockNuxtContent = MockNuxtContent(Array(10).fill({}))
      const wrapper = mount(ListStories, defaultOptionsFactory({
        propsData: {
          limit: 3
        },
        mocks: {
          $content: mockNuxtContent
        }
      }))
      await (ListStories as any).options.fetch.call(wrapper.vm)
      expect(mockNuxtContent().limit).toHaveBeenCalledWith(3)
    })
  })

  describe('Randomize: true', (): void => {
    test('shall { metaPostStories } be Array fetch output Object', async (): Promise<void> => {
      const wrapper = mount(ListStories, defaultOptionsFactory({
        mocks: {
          $content: MockNuxtContent({})
        },
        propsData: {
          randomize: true
        }
      }))
      await (ListStories as any).options.fetch.call(wrapper.vm)
      expect(Array.isArray((wrapper.vm as any).metaPostStories)).toBeTruthy()
    })

    test('shall { metaPostStories } be Array.length = 3 when limit = 3', async (): Promise<void> => {
      const wrapper = mount(ListStories, defaultOptionsFactory({
        propsData: {
          limit: 3,
          randomize: true
        }
      }))
      await (ListStories as any).options.fetch.call(wrapper.vm)
      expect((wrapper.vm as any).metaPostStories).toHaveLength(3)
    })
  })
})
