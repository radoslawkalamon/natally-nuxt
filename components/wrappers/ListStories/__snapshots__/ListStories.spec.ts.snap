// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wrappers / List Stories Unit Randomize: false shall { metaPostStories } match snapshot 1`] = `
[
  DTOMetaPostStory {},
  DTOMetaPostStory {},
  DTOMetaPostStory {},
  DTOMetaPostStory {},
  DTOMetaPostStory {},
  DTOMetaPostStory {},
  DTOMetaPostStory {},
  DTOMetaPostStory {},
  DTOMetaPostStory {},
  DTOMetaPostStory {},
]
`;

exports[`Wrappers / List Stories Unit Randomize: true shall { metaPostStories } match snapshot 1`] = `
[
  DTOMetaPostStory {},
  DTOMetaPostStory {},
  DTOMetaPostStory {},
  DTOMetaPostStory {},
  DTOMetaPostStory {},
  DTOMetaPostStory {},
  DTOMetaPostStory {},
  DTOMetaPostStory {},
  DTOMetaPostStory {},
  DTOMetaPostStory {},
]
`;

exports[`Wrappers / List Stories Unit shall pass unit sanity test 1`] = `"<div class="list-stories">This is slot "default" test.</div>"`;
